### ISSUE TITLE
Leaked scalars

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9752]
Migrated from [rt.perl.org#66108](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=66108) (status was 'resolved')

Searchable as RT66108$

**Comment by p5pRT** [comment]
### From @nwc10

Jerry Hedden mailed p5p in 1ff86f510805081400l1fd9b9das47034600aee20390@&#8203;mail\.gmail\.com
http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2008\-05/msg00254\.html

Dave notes&#8203;:

a regression since 5\.10\.0 \- present in B\,M

**Comment by p5pRT** [comment]
### From @jimc

On Thu May 28 08&#8203;:09&#8203;:27 2009\, nicholas wrote&#8203;:
> Jerry Hedden mailed p5p in
> 1ff86f510805081400l1fd9b9das47034600aee20390@&#8203;mail\.gmail\.com
> http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2008\-
> 05/msg00254\.html
> 
> Dave notes&#8203;:
> 
> a regression since 5\.10\.0 \- present in B\,M

another data point\.

I ran DaveM's simpler code using \-Dm\, 
this coaxes a segfault out of perl\.
Hopefully its related \(it has the clone Dave cited


gdb \./perl \-\- \-Dm \-Ilib \.\./bisect\.pl

\#0  Perl\_sv\_setpvn \(my\_perl=0x8364958\, sv=0xabababab\, ptr=0x82579ea ""\,
len=0\) at sv\.c&#8203;:4296
\#1  0x08171529 in Perl\_sv\_vsetpvfn \(my\_perl=0x8364958\, sv=0xabababab\,
pat=0x825786c "%s"\, patlen=2\, 
&nbsp;    args=0xbfffed14\, svargs=0x0\, svmax=0\, maybe\_tainted=0x0\) at sv\.c&#8203;:9114
\#2  0x080fcfe3 in Perl\_vmess \(my\_perl=0x8364958\, pat=0x825786c "%s"\,
args=0xbfffed14\) at util\.c&#8203;:1203
\#3  0x080fdff5 in S\_vdie\_croak\_common \(my\_perl=0x8364958\, 
&nbsp;    pat=0xbfffed14 "$���H���\\207S\\032\\blx%\\b�\\026\-\\bj\\033'\\b"\,
args=0x4e007325\, msglen=0xbfffecdc\, 
&nbsp;    utf8=0xbfffecd8\) at util\.c&#8203;:1345
\#4  0x080fe48c in Perl\_vcroak \(my\_perl=0x8364958\, pat=0x825786c "%s"\,
args=0xbfffed14\) at util\.c&#8203;:1417
\#5  0x08102b1f in Perl\_croak\_nocontext \(pat=0x825786c "%s"\) at util\.c&#8203;:1438
\#6  0x081a5387 in Perl\_tmps\_grow \(my\_perl=0x8364958\, n=1\) at scope\.c&#8203;:146
\#7  0x081394f0 in Perl\_sv\_2mortal \(my\_perl=0x8364958\, sv=0x837a874\) at
sv\.c&#8203;:7590
\#8  0x081151cd in Perl\_mg\_get \(my\_perl=0x8364958\, sv=0x837a874\) at mg\.c&#8203;:205
\#9  0x0814963f in Perl\_sv\_2pv\_flags \(my\_perl=0x8364958\, sv=0x837a874\,
lp=0xbfffee78\, 
&nbsp;    flags=\<value optimized out>\) at sv\.c&#8203;:2741
\#10 0x0815f318 in Perl\_sv\_pvn\_force\_flags \(my\_perl=0x8364958\,
sv=0x837a874\, lp=0xbfffef9c\, flags=2\) at sv\.c&#8203;:8302
\#11 0x081434b2 in Perl\_sv\_vcatpvfn \(my\_perl=0x8364958\, sv=0x837a874\, 
&nbsp;    pat=0x8280324 "0x%lx&#8203;: \(%05ld\) malloc %ld bytes\\n"\, patlen=32\,
args=0xbffff0b8\, svargs=0x0\, svmax=0\, 
&nbsp;    maybe\_tainted=0x0\) at sv\.c&#8203;:9217
\#12 0x08171911 in Perl\_vnewSVpvf \(my\_perl=0x8364958\, pat=0x8280324
"0x%lx&#8203;: \(%05ld\) malloc %ld bytes\\n"\, 
&nbsp;    args=0xbffff0b8\) at sv\.c&#8203;:7817
\#13 0x08228e7e in PerlIO\_vprintf \(f=0xabababb7\, fmt=0x8280324 "0x%lx&#8203;:
\(%05ld\) malloc %ld bytes\\n"\, 
&nbsp;    ap=0xbffff0d8 " �/\\b����\\234\\001"\) at perlio\.c&#8203;:5120
\#14 0x0822901f in PerlIO\_printf \(f=0xabababb7\, fmt=0x8280324 "0x%lx&#8203;:
\(%05ld\) malloc %ld bytes\\n"\)
&nbsp;    at perlio\.c&#8203;:5135
\#15 0x081033a2 in Perl\_safesysmalloc \(size=\<value optimized out>\) at
util\.c&#8203;:96
\#16 0x08116667 in Perl\_reentrant\_init \(my\_perl=0x8364958\) at reentr\.c&#8203;:148
\#17 0x081639c0 in perl\_clone \(proto\_perl=0x82db008\, flags=2\) at sv\.c&#8203;:11838
\#18 0x007a8a3e in S\_ithread\_create \(\) at threads\.xs&#8203;:758
\#19 XS\_threads\_create \(my\_perl=0x82db008\, cv=0x83415fc\) at threads\.xs&#8203;:1052
\#20 0x08128001 in Perl\_pp\_entersub \(my\_perl=0x82db008\) at pp\_hot\.c&#8203;:2878
\#21 0x080f1cc7 in Perl\_runops\_debug \(my\_perl=0x82db008\) at dump\.c&#8203;:1981
\#22 0x0807f523 in S\_run\_body \(\) at perl\.c&#8203;:2275
\#23 perl\_run \(my\_perl=0x82db008\) at perl\.c&#8203;:2201
\#24 0x0805fb25 in main \(argc=4\, argv=0xbffff4b4\, env=0xbffff4c8\) at
perlmain\.c&#8203;:117
\(gdb\) 


BTW\, adding an \_\_END\_\_ does NOT give me a 3rd attempt to free\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

Bisect of Dave's test case&#8203;:

Running the prog '/tmp/rt\-66108\-2\.pl' for installed\-perls/perl/pSgz7tX/
perl\-5\.9\.4@&#8203;28770/bin/perl and installed\-perls/perl/pz0rj1d/perl\-
5\.9\.4@&#8203;28771/bin/perl
\-\-\-\-Program\-\-\-\-
\#\!/usr/bin/perl \-l

my $out = qx\#$^X /tmp/rt\-66108\.pl 2>&1\#;
$out =~ s/0x\.\.\.\.\.\.\./0xFFFFFFF/g;
print $out;

\-\-\-\-Output of \.\.\./pSgz7tX/perl\-5\.9\.4@&#8203;28770/bin/perl\-\-\-\-


\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\./pz0rj1d/perl\-5\.9\.4@&#8203;28771/bin/perl\-\-\-\-
Attempt to free unreferenced scalar&#8203;: SV 0xFFFFFFF\, Perl interpreter&#8203;: 
0xFFFFFFF at /tmp/rt\-66108\.pl line 24\.
Attempt to free unreferenced scalar&#8203;: SV 0xFFFFFFF\, Perl interpreter&#8203;: 
0xFFFFFFF at /tmp/rt\-66108\.pl line 24\.
Scalars leaked&#8203;: 2


\-\-\-\-EOF \($?='0'\)\-\-\-\-


http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/
78c72037c327e2cd8ede6cf098324435a670ca67
author    Nicholas Clark \<nick@&#8203;ccl4\.org>
&nbsp;    Thu\, 31 Aug 2006 09&#8203;:05&#8203;:50 \+0000 \(09&#8203;:05 \+0000\)
committer    Nicholas Clark \<nick@&#8203;ccl4\.org>
&nbsp;    Thu\, 31 Aug 2006 09&#8203;:05&#8203;:50 \+0000 \(09&#8203;:05 \+0000\)
commit    78c72037c327e2cd8ede6cf098324435a670ca67
tree    a16b884a1b38e00a54fe96eb7fe890408d6abb13    tree | snapshot
parent    2e0df0e80ba5355d269a0006da239be95c70765d    commit | diff

Change the generation of \{\} and \[\] from 3 ops to 1\, and avoid 1 mortal
on the tempstack\, by augmenting pp\_anonlist and pp\_anonhash to accept
OPf\_SPECIAL to mean "return a reference to the aggregate" on the stack
rather than the aggregate itself\.

p4raw\-id&#8203;: //depot/perl@&#8203;28771

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Bisect of Dave's test case&#8203;:

Running the prog '/tmp/rt\-66108\-2\.pl' for installed\-perls/perl/pSgz7tX/
perl\-5\.9\.4@&#8203;28770/bin/perl and installed\-perls/perl/pz0rj1d/perl\-
5\.9\.4@&#8203;28771/bin/perl
\-\-\-\-Program\-\-\-\-
\#\!/usr/bin/perl \-l

my $out = qx\#$^X /tmp/rt\-66108\.pl 2>&1\#;
$out =~ s/0x\.\.\.\.\.\.\./0xFFFFFFF/g;
print $out;

\-\-\-\-Output of \.\.\./pSgz7tX/perl\-5\.9\.4@&#8203;28770/bin/perl\-\-\-\-


\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\./pz0rj1d/perl\-5\.9\.4@&#8203;28771/bin/perl\-\-\-\-
Attempt to free unreferenced scalar&#8203;: SV 0xFFFFFFF\, Perl interpreter&#8203;: 
0xFFFFFFF at /tmp/rt\-66108\.pl line 24\.
Attempt to free unreferenced scalar&#8203;: SV 0xFFFFFFF\, Perl interpreter&#8203;: 
0xFFFFFFF at /tmp/rt\-66108\.pl line 24\.
Scalars leaked&#8203;: 2


\-\-\-\-EOF \($?='0'\)\-\-\-\-


http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/
78c72037c327e2cd8ede6cf098324435a670ca67
author    Nicholas Clark \<nick@&#8203;ccl4\.org>
&nbsp;    Thu\, 31 Aug 2006 09&#8203;:05&#8203;:50 \+0000 \(09&#8203;:05 \+0000\)
committer    Nicholas Clark \<nick@&#8203;ccl4\.org>
&nbsp;    Thu\, 31 Aug 2006 09&#8203;:05&#8203;:50 \+0000 \(09&#8203;:05 \+0000\)
commit    78c72037c327e2cd8ede6cf098324435a670ca67
tree    a16b884a1b38e00a54fe96eb7fe890408d6abb13    tree | snapshot
parent    2e0df0e80ba5355d269a0006da239be95c70765d    commit | diff

Change the generation of \{\} and \[\] from 3 ops to 1\, and avoid 1 mortal
on the tempstack\, by augmenting pp\_anonlist and pp\_anonhash to accept
OPf\_SPECIAL to mean "return a reference to the aggregate" on the stack
rather than the aggregate itself\.

p4raw\-id&#8203;: //depot/perl@&#8203;28771

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, May 28\, 2009 at 08&#8203;:09&#8203;:27AM \-0700\, Nicholas Clark wrote&#8203;:
> \# New Ticket Created by  Nicholas Clark 
> \# Please include the string&#8203;:  \[perl \#66108\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=66108 >
> 
> 
> Jerry Hedden mailed p5p in 1ff86f510805081400l1fd9b9das47034600aee20390@&#8203;mail\.gmail\.com
> http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2008\-05/msg00254\.html
> 
> Dave notes&#8203;:
> 
> a regression since 5\.10\.0 \- present in B\,M

Now fixed&#8203;:

commit 7fa38291524c327a3cb23bfe94979e1537743cac
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Fri Jan 8 23&#8203;:31&#8203;:45 2010 \+0000
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Fri Jan 8 23&#8203;:31&#8203;:45 2010 \+0000

&nbsp;    fix for \[perl \#66108\] Leaked scalars
&nbsp;    
&nbsp;    @&#8203;DB&#8203;::args is a hack&#8203;: it gets set with non\-refcounted aliases of the
&nbsp;    caller's @&#8203;\_ elements\.
&nbsp;    
&nbsp;    Once the sub that ran caller\(\) has exited\, @&#8203;DB&#8203;::args will contain garbage&#8203;:
&nbsp;    elements will be SVs that have been freed\, re\-assigned etc\.
&nbsp;    
&nbsp;    So as a minimum\, when cloning an interpreter\, skip cloning @&#8203;DB&#8203;::args\.


Affected files \.\.\.
&nbsp;    
&nbsp;    M    sv\.c

Differences \.\.\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/sv.c b/sv.c
index 38a9140..c2757d6 100644
--- a/sv.c
+++ b/sv.c
@@ -11913,6 +11913,10 @@ perl_clone_using(PerlInterpreter *proto_perl, UV flags,
     SvNV_set(&PL_sv_yes, 1);
     ptr_table_store(PL_ptr_table, &proto_perl->Isv_yes, &PL_sv_yes);
 
+    /* dbargs array probably holds garbage; give the child a clean array */
+    PL_dbargs		= newAV();
+    ptr_table_store(PL_ptr_table, proto_perl->Idbargs, PL_dbargs);
+
     /* create (a non-shared!) shared string table */
     PL_strtab		= newHV();
     HvSHAREKEYS_off(PL_strtab);
@@ -12039,7 +12043,6 @@ perl_clone_using(PerlInterpreter *proto_perl, UV flags,
     PL_DBsingle		= sv_dup(proto_perl->IDBsingle, param);
     PL_DBtrace		= sv_dup(proto_perl->IDBtrace, param);
     PL_DBsignal		= sv_dup(proto_perl->IDBsignal, param);
-    PL_dbargs		= av_dup(proto_perl->Idbargs, param);
 
     /* symbol tables */
     PL_defstash		= hv_dup_inc(proto_perl->Idefstash, param);


-- 
```
</details>

Lady Nancy Astor&#8203;: If you were my husband\, I would flavour your coffee
with poison\.
Churchill&#8203;: Madam \- if I were your husband\, I would drink it\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
